lib/command/src/Obelisk/Command/Run.hs

```diff
@@ -105,38 +107,50 @@ runWatch = do
   pkgs <- getLocalPkgs
   withGhciScript pkgs $ \dotGhciPath -> runGhcid dotGhciPath Nothing
 
+data GuessPackageFileError = GuessPackageFileError_Ambiguous [FilePath] | GuessPackageFileError_NotFound
+  deriving (Eq, Ord, Show)
+instance Exception GuessPackageFileError
+
+-- | Given a directory, try to guess what the appropriate @.cabal@ or @package.yaml@ file is for the package.
+guessCabalPackageFile
+  :: MonadObelisk m
+  => FilePath -- ^ Directory to search for cabal package
+  -> m (Either GuessPackageFileError (Either FilePath FilePath))
+  -- ^ 'Right' 'Left' for hpack package, 'Right' 'Right' for cabal package; 'FilePath' is relative to given directory.
+guessCabalPackageFile dir = do
+  candidates <- liftIO $
+        filterM doesFileExist
+    =<< filter (liftA2 (||) (== Hpack.packageConfig) (".cabal" `L.isSuffixOf`))
+    <$> listDirectory dir
+  pure $ case L.partition (== Hpack.packageConfig) candidates of
+    ([hpack], _) -> Right $ Left hpack
+    ([], [cabal]) -> Right $ Right cabal
+    ([], []) -> Left GuessPackageFileError_NotFound
+    (hpacks, cabals) -> Left $ GuessPackageFileError_Ambiguous $ hpacks <> cabals
+
 -- | Relative paths to local packages of an obelisk project
 -- TODO a way to query this
-getLocalPkgs :: Applicative f => f [(FilePath, Maybe String)]
-getLocalPkgs = pure [("backend", Nothing), ("common", Nothing), ("frontend", Nothing)]
+getLocalPkgs :: Applicative f => f [FilePath]
+getLocalPkgs = pure ["backend", "common", "frontend"]
 
 parseCabalPackage
   :: (MonadObelisk m)
   => FilePath -- ^ package directory
-  -> Maybe String -- ^ package name, defaults to directory
   -> m (Maybe CabalPackageInfo)
-parseCabalPackage dir name = do
-  let cabalFp = dir </> (fromMaybe (takeBaseName dir) name <> ".cabal")
-      hpackFp = dir </> "package.yaml"
-  hasCabal <- liftIO $ doesFileExist cabalFp
-  hasHpack <- liftIO $ doesFileExist hpackFp
-
-  mCabalContents <- if hasCabal
-    then Just <$> liftIO (readUTF8File cabalFp)
-    else if hasHpack
-      then do
-        let decodeOptions = DecodeOptions (ProgramName "ob") hpackFp Nothing decodeYaml
-        liftIO (readPackageConfig decodeOptions) >>= \case
-          Left err -> do
-            putLog Error $ T.pack $ "Failed to parse " <> hpackFp <> ": " <> err
-            return Nothing
-          Right (DecodeResult hpackPackage _ _ _) -> do
-            return $ Just $ renderPackage [] hpackPackage
-      else do
-        putLog Error $ T.pack "Found neither cabal nor hpack file"
-        return Nothing
+parseCabalPackage dir = do
```

Comment:
Maybe it would be better to have the bulk of this function run inside `ExceptT (NonEmpty Text)`, ending with

```haskell 
\case
  Left errs -> for errs (putLog Error) *> pure Nothing
  Right info -> pure info
```
I feel that any failure here should provide a user-facing reason, rather than be allowed to simply return `Nothing` without logs of any kind (e.g. #597).

Reply:
At this point the function does provide a user facing reason in any place that it returns `Nothing`.

Reply:
I see your point though.

Reply:
Done.